### PR TITLE
UF-8560: Adjusted API tests

### PR DESCRIPTION
**Pull request by josc04**
- Modified API success tests to use Spring WebTestClient
- Replaced old validation tests with API failure tests using Spring WebTestClient
- Extracted mapping logic from resource classes to mapper class
- Updated documentation for API resources that throws bad request
- Updated to latest version of Dept-44

### CODE REVIEW

File: src/test/java/se/sundsvall/messaging/api/MessageResourceDigitalMailFailureTests.java [hunk 1]
```diff
@@ -0,0 +1,389 @@
...
+
+		verifyNoInteractions(mockMessageService, mockEventDispatcher);
+	}
+
+	@ParameterizedTest
+	@ValueSource(strings = " ")
+	@NullAndEmptySource
+	void shouldFailWithInvalidExternalReferenceKey(String key) {
+		// Arrange
+		final var externalReference = validRequest.party().externalReferences().get(0);
```
Comment (mha109) [inline src/test/java/se/sundsvall/messaging/api/MessageResourceDigitalMailFailureTests.java:110]: `getFirst()`?

File: src/test/java/se/sundsvall/messaging/api/MessageResourceDigitalInvoiceTests.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+import org.springframework.test.web.reactive.server.WebTestClient;
+
+import se.sundsvall.messaging.Application;
+import se.sundsvall.messaging.api.model.response.MessageResult;
+import se.sundsvall.messaging.model.InternalDeliveryResult;
+import se.sundsvall.messaging.service.MessageEventDispatcher;
+import se.sundsvall.messaging.service.MessageService;
+
+@SpringBootTest(classes = Application.class, webEnvironment = RANDOM_PORT)
+@ActiveProfiles("junit")
```
Comment (mha109) [inline src/test/java/se/sundsvall/messaging/api/MessageResourceDigitalInvoiceTests.java:30]: There's a `@UnitTest`-annotation, maybe use it for concistency?

File: src/test/java/se/sundsvall/messaging/api/MessageResourceEmailBatchTests.java [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+	@MockBean
+	private MessageEventDispatcher mockEventDispatcher;
+
+	@Autowired
+	private WebTestClient webTestClient;
+
+	@ParameterizedTest
+	@MethodSource("requestProvider")
+	void sendBatch(String senderName, String partyId) {
```
Comment (mha109) [inline src/test/java/se/sundsvall/messaging/api/MessageResourceEmailBatchTests.java:65]: Parameters are not used?

File: src/main/java/se/sundsvall/messaging/api/StatusAndHistoryResource.java [hunk 4]
```diff
@@ -68,328 +60,115 @@ class StatusAndHistoryResource {
...
-				content = @Content(schema = @Schema(implementation = Problem.class))
-			)
-		}
-	)
-	@GetMapping(
-		value = BATCH_STATUS_PATH,
-		produces = {APPLICATION_JSON_VALUE, APPLICATION_PROBLEM_JSON_VALUE}
-	)
+	@Operation(summary = "Get the status for a message batch, its messages and their deliveries", responses = {
+		@ApiResponse(responseCode = "200", description = "Successful Operation", content = @Content(schema = @Schema(implementation = MessageBatchResult.class))),
```
Comment (foobar-1337) [inline src/main/java/se/sundsvall/messaging/api/StatusAndHistoryResource.java:80]: useReturnTypeSchema?

File: src/main/java/se/sundsvall/messaging/api/StatusAndHistoryResource.java [hunk 5]
```diff
@@ -68,328 +60,115 @@ class StatusAndHistoryResource {
...
-				content = @Content(schema = @Schema(implementation = Problem.class))
-			)
-		}
-	)
-	@GetMapping(
-		value = "/conversation-history/{partyId}",
-		produces = {APPLICATION_JSON_VALUE, APPLICATION_PROBLEM_JSON_VALUE}
-	)
+	@Operation(summary = "Get the entire conversation history for a given party", responses = {
+		@ApiResponse(responseCode = "200", description = "Successful Operation", content = @Content(array = @ArraySchema(schema = @Schema(implementation = HistoryResponse.class)))),
```
Comment (foobar-1337) [inline src/main/java/se/sundsvall/messaging/api/StatusAndHistoryResource.java:64]: useReturnTypeSchema?

File: src/main/java/se/sundsvall/messaging/api/StatusAndHistoryResource.java [hunk 6]
```diff
@@ -68,328 +60,115 @@ class StatusAndHistoryResource {
...
-				content = @Content(schema = @Schema(implementation = Problem.class))
-			)
-		}
-	)
-	@GetMapping(
-		value = MESSAGE_STATUS_PATH,
-		produces = {APPLICATION_JSON_VALUE, APPLICATION_PROBLEM_JSON_VALUE}
-	)
+	@Operation(summary = "Get the status for a single message and its deliveries", responses = {
+		@ApiResponse(responseCode = "200", description = "Successful Operation", content = @Content(schema = @Schema(implementation = MessageResult.class))),
```
Comment (foobar-1337) [inline src/main/java/se/sundsvall/messaging/api/StatusAndHistoryResource.java:96]: useReturnTypeSchema?

File: src/main/java/se/sundsvall/messaging/api/StatusAndHistoryResource.java [hunk 7]
```diff
@@ -68,328 +60,115 @@ class StatusAndHistoryResource {
...
-				content = @Content(schema = @Schema(implementation = Problem.class))
-			)
-		}
-	)
-	@GetMapping(
-		value = DELIVERY_STATUS_PATH,
-		produces = {APPLICATION_JSON_VALUE, APPLICATION_PROBLEM_JSON_VALUE}
-	)
+	@Operation(summary = "Get the status for a single delivery", responses = {
+		@ApiResponse(responseCode = "200", description = "Successful Operation", content = @Content(schema = @Schema(implementation = DeliveryResult.class))),
```
Comment (foobar-1337) [inline src/main/java/se/sundsvall/messaging/api/StatusAndHistoryResource.java:112]: useReturnTypeSchema?

File: src/main/java/se/sundsvall/messaging/api/StatusAndHistoryResource.java [hunk 8]
```diff
@@ -68,328 +60,115 @@ class StatusAndHistoryResource {
...
-				content = @Content(schema = @Schema(implementation = Problem.class))
-			)
-		}
-	)
-	@GetMapping(
-		value = "/message/{messageId}",
-		produces = {APPLICATION_JSON_VALUE, APPLICATION_PROBLEM_JSON_VALUE}
-	)
+	@Operation(summary = "Get a message and all its deliveries", responses = {
+		@ApiResponse(responseCode = "200", description = "Successful Operation", content = @Content(array = @ArraySchema(schema = @Schema(implementation = HistoryResponse.class)))),
```
Comment (foobar-1337) [inline src/main/java/se/sundsvall/messaging/api/StatusAndHistoryResource.java:128]: useReturnTypeSchema?

File: src/main/java/se/sundsvall/messaging/api/StatusAndHistoryResource.java [hunk 9]
```diff
@@ -68,328 +60,115 @@ class StatusAndHistoryResource {
...
-		@Parameter(description = "From-date (inclusive). Format: yyyy-MM-dd (ISO8601)") final LocalDate from,
-
-		@RequestParam(name = "to", required = false)
-		@DateTimeFormat(iso = DateTimeFormat.ISO.DATE)
-		@Parameter(description = "To-date (inclusive). Format: yyyy-MM-dd (ISO8601)") final LocalDate to) {
-		var stats = statisticsService.getStatistics(messageType, from, to);
-
-		return ResponseEntity.ok(stats);
+	@Operation(summary = "Get delivery statistics", responses = {
+		@ApiResponse(responseCode = "200", description = "Successful Operation", content = @Content(array = @ArraySchema(schema = @Schema(implementation = Statistics.class)))),
```
Comment (foobar-1337) [inline src/main/java/se/sundsvall/messaging/api/StatusAndHistoryResource.java:147]: useReturnTypeSchema?

File: src/test/java/se/sundsvall/messaging/api/MessageResourceEmailBatchTests.java (hunk 3 above)
Comment (josc04) [reply src/test/java/se/sundsvall/messaging/api/MessageResourceEmailBatchTests.java:65]: Will remove, they serve no purpose in this test

File: src/test/java/se/sundsvall/messaging/api/MessageResourceDigitalInvoiceTests.java [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+class MessageResourceDigitalInvoiceTests {
```
[the remark this answers is not in the document]
Comment (josc04) [reply src/test/java/se/sundsvall/messaging/api/MessageResourceDigitalInvoiceTests.java:31]: This service will have a reboot to "make it great again". Tests will be renamed in that story. I.e. I will not adjust testclass names in this PR.

[the remark this answers is not in the document]
Comment (foobar-1337) [reply src/test/java/se/sundsvall/messaging/api/MessageResourceDigitalInvoiceTests.java:31]: Boyscout?

[the remark this answers is not in the document]
Comment (josc04) [reply src/test/java/se/sundsvall/messaging/api/MessageResourceDigitalInvoiceTests.java:31]: I knew you were gonna say that. Well, you know the drill... hawk tuah...

**Comment by josc04** [comment]
Fixed review comments

File: src/integration-test/java/openapi/OpenApiSpecificationIT.java [hunk 11]
```diff
@@ -46,7 +44,7 @@ class OpenApiSpecificationIT {
 	void compareOpenApiSpecifications() {
 		final var existingOpenApiSpecification = asString(openApiResource);
 		final var currentOpenApiSpecification = getCurrentOpenApiSpecification();
-
+		System.err.println(currentOpenApiSpecification);
```
Comment (foobar-1337) [inline src/integration-test/java/openapi/OpenApiSpecificationIT.java]: Remove?

**Review by foobar-1337 — APPROVED** [review]

**Pull request merged**